fea-rs/src/parse/source.rs

```diff
@@ -183,69 +268,53 @@ impl SourceMap {
 }
 
 impl SourceList {
-    pub(crate) fn new(root: Source, project_root: Option<PathBuf>) -> Self {
-        // if root has no source it means we're testing, and project_root can
-        // be empty? (future me: if this caused a bug, I'm sorry.
-        // it looked harmless.)
-        let project_root = project_root.unwrap_or_else(|| match root.path() {
-            // if path is real, it should exist and have a parent
-            Some(path) => path.parent().unwrap().to_owned(),
-            None => PathBuf::new(),
-        });
-        let root_id = root.id;
-        let canonical = match root.path() {
-            Some(path) => util::paths::CanonicalPath::from_path(path).unwrap(),
-            None => util::paths::CanonicalPath::fake(format!("path/to/{}.fea", root_id.0)),
-        };
-
-        let mut myself = SourceList {
-            project_root,
-            root_id,
+    pub(crate) fn new(resolver: impl SourceResolver + 'static) -> Self {
```

Comment:
I think I might have expected `Box<dyn SourceResolver>`, I'm curious why this?

Reply:
This is about ergonomics; by taking `impl trait` and then boxing inside, we don't require the user to explicitly call `Box::new(my_thing)`, which is annoying (particularly in the case where you're passing a closure).